### PR TITLE
feat(provider): add swap cache TTL billing option

### PR DESCRIPTION
**Pull request by miraserver**
## Summary

Adds a per-provider **Swap Cache TTL Billing** toggle that inverts the 1-hour / 5-minute cost calculation. When enabled on a provider, requests routed through it apply 5-minute pricing where 1-hour pricing would normally apply, and vice versa. This allows operators to reflect actual upstream billing models that differ from the default.

## Problem

Some providers report `cache_ttl=1h` in their API responses but actually bill at the 5-minute cache rate. The existing differentiated billing (Issue #278) correctly distinguishes between 5min and 1h cache tokens, but cannot handle cases where the TTL indicator doesn't match the actual billing model. This PR adds a per-provider override to handle such discrepancies.

**Related Issues:**
- Closes #278 - Added differentiated 5min/1hr cache billing support (foundation for this PR)
- Related to #277 - Original feature request for cache time differentiation

### Database (2 migrations)
- `0069`: adds `swap_cache_ttl_billing` boolean column to `providers` table (default `false`)
- `0070`: adds `swap_cache_ttl_applied` boolean column to `message_request` table (default `false`)

### Provider Settings
- New toggle in the Routing section of the provider form
- Displayed on provider list items and vendor-keys compact view
- Persisted through provider create/update actions

### Proxy Pipeline
- `response-handler.ts`: extracted `applySwapCacheTtlBilling()` utility; applies swap at data-entry point so cost, badge, and Langfuse metrics stay consistent
- `error-handler.ts`: propagates `swapCacheTtlApplied` flag on error responses
- Clones `usageMetrics` before mutation to prevent side-effects

### Logs UI
- `cacheTtlApplied` badge in usage-logs-table shows a swap indicator when billing was inverted
- Error details dialog (SummaryTab, MetadataTab) displays the swap flag
- Reordered swap toggle placement and added missing `DialogTitle` for a11y

### Bugfix
- `provider-manager.tsx`: fixed duplicate `"default"` key in provider group tabs when a provider explicitly has `groupTag = "default"`

### i18n (5 languages: en, ja, ru, zh-CN, zh-TW)
- Added translation keys for swap toggle label, description, and badge indicator in `dashboard.json` and `settings/providers/form/sections.json`

## i18n Audit Evidence

- `bun run i18n:audit-placeholders` — all placeholders consistent across languages
- `bun run i18n:audit-messages-no-emoji:fail` — no emoji in any message file
- `settings-split-guards.test.ts` and `settings-index-modules-load.test.ts` — pass

## Test Coverage

- **20 feature tests** pass (cost calculation swap logic, logs UI rendering, i18n guards)
- `tests/unit/lib/cost-calculation-swap-cache-ttl.test.ts` — 314-line dedicated test suite covering:
  - Happy path: swap applied correctly for both 1h->5min and 5min->1h directions
  - Edge cases: zero usage, missing metrics, provider flag off
  - Mutation safety: original usageMetrics object not modified
- `usage-logs-table.test.tsx` and `virtualized-logs-table.test.tsx` — badge rendering with swap indicator

## Live Testing

- 12 requests through proxy pipeline, 0 errors
- Swap logic verified correct in DB logs (swapCacheTtlApplied = true only for enabled providers)

## PR Checklist

- [x] Base branch = `dev`
- [x] Pre-commit checks pass (build, lint, typecheck, test)
- [x] i18n audit clean (no emoji, placeholders OK, all 5 languages)
- [x] No conflicts with dev
- [x] Migrations generated via `bun run db:generate`, not manually

---
*Description enhanced by Claude AI*

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds a per-provider **Swap Cache TTL Billing** toggle that inverts the 5-minute and 1-hour cache token cost buckets. The feature addresses cases where upstream providers report a cache TTL that doesn't match their actual billing model.

- **Core logic**: `applySwapCacheTtlBilling()` swaps the 5m/1h cache token buckets and inverts the `cache_ttl` indicator. `normalizeUsageWithSwap()` wraps this with clone-before-mutate safety and session fallback TTL resolution. Applied consistently across non-stream, stream, and Gemini passthrough paths.
- **Refactored `finalizeRequestStats`** to return the normalized `UsageMetrics` so Langfuse traces receive swap-aware usage data instead of re-parsing from raw response text.
- **Database**: Two clean migrations add `swap_cache_ttl_billing` (providers) and `swap_cache_ttl_applied` (message_request) columns.
- **UI**: Toggle in provider routing settings; amber-styled swap indicator badge in logs tables and error detail dialogs.
- **Bugfix**: Fixes duplicate "default" key in provider group tabs when a provider explicitly has `groupTag = "default"`.
- **A11y**: Adds missing `DialogTitle` (via `VisuallyHidden`) to multiple provider dialogs.
- **i18n**: All 5 languages (en, ja, ru, zh-CN, zh-TW) updated with new translation keys.
- **Tests**: 330-line dedicated test suite covers swap logic, edge cases, mutation safety, and idempotency. UI tests cover badge rendering.
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

- This PR is safe to merge with low risk — the swap logic is well-tested and the clone-before-mutate pattern prevents side-effects.
- Score of 4 reflects: thorough test coverage (330-line dedicated test suite + UI tests), consistent application across all code paths (non-stream, stream, Gemini passthrough, error, failure), clean database migrations, complete i18n coverage, and a well-thought-out clone-before-mutate pattern. The only concern is the exported in-place-mutating `applySwapCacheTtlBilling` function which could invite misuse by future callers, but current usage is correct.
- `src/app/v1/_lib/proxy/response-handler.ts` deserves closest attention as it contains the core swap logic and touches multiple response handling code paths.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/app/v1/_lib/proxy/response-handler.ts | Core swap logic added via `applySwapCacheTtlBilling` and `normalizeUsageWithSwap`. Refactored `finalizeRequestStats` to return normalized usage for Langfuse trace consistency. Clone-before-mutate pattern prevents double-swap. Well-structured overall with one minor naming concern. |
| src/app/v1/_lib/proxy/error-handler.ts | Single-line addition to propagate `swapCacheTtlApplied` flag on error persistence. Clean and correct. |
| src/drizzle/schema.ts | Two new boolean columns added: `swap_cache_ttl_billing` on providers (NOT NULL, default false) and `swap_cache_ttl_applied` on message_request (default false). Schema matches migrations. |
| src/repository/provider.ts | Added `swapCacheTtlBilling` to all provider query projections and create/update operations. Consistently applied across all 6 query locations. |
| src/app/[locale]/dashboard/logs/_components/usage-logs-table.tsx | Added swap indicator badge with amber styling when `swapCacheTtlApplied` is true. Passes flag through to error details dialog. |
| src/app/[locale]/dashboard/logs/_components/virtualized-logs-table.tsx | Added swap indicator badge with conditional amber/blue styling. Consistent with usage-logs-table changes. |
| src/app/[locale]/settings/providers/_components/forms/provider-form/sections/routing-section.tsx | Added Swap Cache TTL Billing toggle in the routing section. Placed correctly before the Cache TTL preference input. |
| src/app/[locale]/settings/providers/_components/provider-manager.tsx | Bugfix: prevents duplicate "default" key in provider group tabs when a provider explicitly has `groupTag = "default"`. Correct logic. |
| tests/unit/lib/cost-calculation-swap-cache-ttl.test.ts | 314-line comprehensive test suite covering swap logic, edge cases (zero/undefined/mixed), mutation safety, and double-swap idempotency. Well-written. |

</details>

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant Client
    participant Proxy as Proxy Pipeline
    participant SwapFn as normalizeUsageWithSwap()
    participant CostCalc as Cost Calculation
    participant DB as Database
    participant Langfuse

    Client->>Proxy: Request (routed to provider)
    Proxy->>Proxy: Receive upstream response
    Proxy->>Proxy: parseUsageFromResponseText()
    
    alt provider.swapCacheTtlBilling = true
        Proxy->>SwapFn: Clone usage + swap 5m↔1h buckets
        SwapFn->>SwapFn: Invert cache_ttl (5m↔1h)
        SwapFn->>SwapFn: Resolve session fallback TTL (also inverted)
        SwapFn-->>Proxy: Normalized UsageMetrics (swapped)
    else provider.swapCacheTtlBilling = false
        Proxy->>SwapFn: Clone usage (no swap)
        SwapFn-->>Proxy: Normalized UsageMetrics (unchanged)
    end

    Proxy->>CostCalc: Calculate cost with normalized usage
    Proxy->>DB: Persist usage + swapCacheTtlApplied flag
    Proxy->>Langfuse: Emit trace with normalized usage
    Proxy-->>Client: Response
```
</details>

<sub>Last reviewed commit: dfbca74</sub>

### CODE REVIEW

File: src/app/[locale]/dashboard/logs/_components/error-details-dialog/components/MetadataTab.tsx [hunk 1]
```diff
@@ -205,8 +206,18 @@ export function MetadataTab({
...
+                    className={cn(
+                      "text-xs",
+                      swapCacheTtlApplied
+                        ? "bg-amber-50 text-amber-700 border-amber-200 dark:bg-amber-950/30 dark:text-amber-300 dark:border-amber-800"
+                        : ""
+                    )}
+                    title={swapCacheTtlApplied ? t("billingDetails.cacheTtlSwapped") : undefined}
+                  >
                     {cacheTtlApplied}
+                    {swapCacheTtlApplied ? " ~" : ""}
```
[the remark this answers is not in the document]
Comment (miraserver) [reply src/app/[locale]/dashboard/logs/_components/error-details-dialog/components/MetadataTab.tsx:220]: The `~` character here is a **visual indicator/symbol**, not translatable text — similar to how icons or mathematical symbols are used in badges. Creating an i18n key that maps to the same `" ~"` across all 5 locales adds indirection without any localization benefit. Declining this suggestion.

File: src/app/[locale]/dashboard/logs/_components/error-details-dialog/components/SummaryTab.tsx [hunk 2]
```diff
@@ -316,8 +317,18 @@ export function SummaryTab({
...
+                      "text-xs",
+                      swapCacheTtlApplied
+                        ? "bg-amber-50 text-amber-700 border-amber-200 dark:bg-amber-950/30 dark:text-amber-300 dark:border-amber-800"
+                        : ""
+                    )}
+                    title={swapCacheTtlApplied ? t("billingDetails.cacheTtlSwapped") : undefined}
+                  >
                     {cacheTtlApplied}
+                    {swapCacheTtlApplied ? " ~" : ""}
                   </Badge>
```
[the remark this answers is not in the document]
Comment (miraserver) [reply src/app/[locale]/dashboard/logs/_components/error-details-dialog/components/SummaryTab.tsx:332]: Same as above — `" ~"` is a visual symbol/indicator, not localizable text. Declining.

File: src/app/[locale]/dashboard/logs/_components/usage-logs-table.tsx [hunk 3]
```diff
@@ -299,8 +299,22 @@ export function UsageLogsTable({
...
+                                  )}
+                                  title={
+                                    log.swapCacheTtlApplied
+                                      ? t("logs.billingDetails.cacheTtlSwapped")
+                                      : undefined
+                                  }
+                                >
                                   {log.cacheTtlApplied}
+                                  {log.swapCacheTtlApplied ? " ~" : ""}
                                 </Badge>
```
[the remark this answers is not in the document]
Comment (miraserver) [reply src/app/[locale]/dashboard/logs/_components/usage-logs-table.tsx:318]: Same as above — `" ~"` is a visual symbol/indicator, not localizable text. Declining.

File: src/app/[locale]/dashboard/logs/_components/usage-logs-table.tsx [hunk 4]
```diff
@@ -299,8 +299,22 @@ export function UsageLogsTable({
...
                             <div className="flex items-center gap-2 w-full cursor-help">
                               {log.cacheTtlApplied ? (
-                                <Badge variant="outline" className="text-[10px] leading-tight px-1">
+                                <Badge
+                                  variant="outline"
+                                  className={cn(
+                                    "text-[10px] leading-tight px-1",
+                                    log.swapCacheTtlApplied
+                                      ? "bg-amber-50 text-amber-700 border-amber-200 dark:bg-amber-950/30 dark:text-amber-300 dark:border-amber-800"
+                                      : ""
```
[the remark this answers is not in the document]
Comment (miraserver) [reply src/app/[locale]/dashboard/logs/_components/usage-logs-table.tsx:308]: The styling difference between the two tables (explicit blue in `virtualized-logs-table` vs default outline in `usage-logs-table`) is **pre-existing** — it was there before this PR. Our changes only added the conditional amber styling for the swapped state in both tables, preserving each table's original non-swapped appearance. Fixing a pre-existing inconsistency is out of scope for this PR.

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 5]
```diff
@@ -2640,6 +2657,59 @@ function adjustUsageForProviderType(
   };
 }
 
+/**
+ * Swap 5m/1h cache buckets and cache_ttl when provider.swapCacheTtlBilling is enabled.
+ * Mutates in-place.
+ */
+export function applySwapCacheTtlBilling(usage: UsageMetrics, swap: boolean | undefined): void {
+  if (!swap) return;
```
[the remark this answers is not in the document]
Comment (miraserver) [reply src/app/v1/_lib/proxy/response-handler.ts:2665]: Fixed in dfbca74 — now using destructuring swap as suggested.

**Comment by miraserver** [comment]
## Review feedback resolution

Pushed 2 commits addressing 6 of 9 bot review findings.

### Accepted & fixed

| # | Finding | Fix |
|---|---------|-----|
| F2 | Explicit `"default"` groupTag creates duplicate tab (CodeRabbit) | `90f83d3` — treat `"default"` tag as `hasDefaultGroup=true` instead of adding to Set |
| F5 | `VisuallyHidden` import in wrong block (CodeRabbit) | `dfbca74` — moved to third-party imports in both files |
| F6 | Terse comment on session fallback TTL inversion (CodeRabbit) | `dfbca74` — expanded to 3-line explanation |
| F7 | `swapCacheTtlApplied` may be `undefined` (CodeRabbit) | `dfbca74` — added `?? false` at all 4 call sites |
| F8 | Temp variable swap is verbose (Gemini) | `dfbca74` — destructuring swap |

### Declined

**F1 — i18n for `" ~"` suffix** (CodeRabbit): The `~` is a visual indicator/symbol, not translatable text. Creating an i18n key that maps to the same `" ~"` across all 5 locales adds indirection without localization benefit.

**F3 — Badge styling inconsistency between tables** (Greptile): The styling difference (explicit blue in `virtualized-logs-table` vs default outline in `usage-logs-table`) is pre-existing. This PR only added conditional amber styling for the swapped state, preserving each table's original appearance.

**F4 — Comment language in `message.ts:259`** (CodeRabbit): Per the project's contribution guidelines, code comments should be in English. The surrounding Chinese comments are a pre-existing convention in this file, but our English comment is consistent with the contribution rules.

### Skipped

**F9 — Direct `normalizeUsageWithSwap` test** (CodeRabbit): Requires a full `ProxySession` mock (including `getCacheTtlResolved()`, provider chain, etc.) for marginal regression coverage. The existing 19-test suite covers the same arithmetic via `applySwapCacheTtlBilling` and bucket routing simulation.

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 6]
```diff
@@ -2640,6 +2657,62 @@ function adjustUsageForProviderType(
...
+  [usage.cache_creation_5m_input_tokens, usage.cache_creation_1h_input_tokens] = [
+    usage.cache_creation_1h_input_tokens,
+    usage.cache_creation_5m_input_tokens,
+  ];
+  if (usage.cache_ttl === "5m") usage.cache_ttl = "1h";
+  else if (usage.cache_ttl === "1h") usage.cache_ttl = "5m";
```
[the remark this answers is not in the document]
Comment (miraserver) [reply src/app/v1/_lib/proxy/response-handler.ts:2671]: The export is intentional — it allows the 19-test suite in `cost-calculation-swap-cache-ttl.test.ts` to exercise the swap arithmetic directly without mocking a full `ProxySession`. The JSDoc explicitly documents `Mutates in-place`, and the only internal caller (`normalizeUsageWithSwap`) clones before invoking it. Making it return a new object would add an unnecessary allocation in the hot proxy path. Keeping as-is.

**Comment by NieiR** [comment]
FYI: this PR introduced half-width parentheses in `messages/zh-TW/dashboard.json`:

```
"cacheTtlSwapped": "計費 TTL (已互換)"
```

This breaks the `zh-tw-dashboard-parentheses` test (which requires fullwidth `（）` only). Fixed in #788 (`2be6f617`): `(已互換)` → `（已互換）`.
